New class DoFVector

guidokanschat opened the pull request:
The new class `DoFVector` introduces better coupling between vectors and finite element spaces.

it serves as a base class for `FEFieldFunction`. The function `DoFVector::sync()` serves to replace reinitializing vector sizes by hand. It is currently implemented for `Vector`, `BlockVector` and  `MGLevelObject` thereof.

tjhei commented on include/deal.II/dofs/dof_vector.h:
```diff
@@ -0,0 +1,243 @@
+  /**
+   * Read-write access to the data vector.
+   *
+   * Read-write access is only possible, if the vector is owned by the
```
Why do you not allow write access in this case?

guidokanschat replied:
Mostly because I did not want to go through the whole "remember constness" issue. If `DoFVector` is only used for reading, you can simply create one from your usual vectors. Better style would be not to cast between regular vectors and `DoFVector`. If we have functions that use it for writing, we indeed need a genuine `DoFVector`, that owns its data.

bangerth replied:
But then I think I still don't understand what the whole thing is there for. If the intent is to create a DoFVector every time one calls a library function, then one still has to list DoFHandler, ConstraintMatrix, and vector. In that case, there is little gained -- just more to write. On the other hand, if this class is intended as a way for people to only store a DoFVector, rather than a plain vector, then it is necessary that they be able to write to it.

tjhei commented:
I have a bunch of questions:
- there is no way to get at the ```ConstraintMatrix``` once you put it in, is that intentional?
- It took me a while to understand what ```sync()``` does (mostly because of the name), maybe it is worth calling it ```reinit()```?
- How does assignment/copy construction of DoFVector objects work?

guidokanschat commented:
@tjhei 

* Access to the `ConstrintMatrix` was forgotten and will be added.
* `reinit()` suggests to me a different semantics, namely changing the whole object. But I'll gladly take a different name than `sync()` for updating the data part to  match the dof handler part
* Assignment and copy constructor are worth discussing. I tend to consider `DoFVector` a replacement of regular vectors in the long run. That means, we should use deep copy. If the `DoFVector` does not own its data, shallow copy seems more reasonable.

tjhei commented on include/deal.II/dofs/dof_vector.h:
```diff
@@ -0,0 +1,243 @@
+   * distributed format, but writing into vectors will most likely
+   * fail without using a constraint matrix from somewhere else.
+   */
+  DoFVector (const DH &dh);
```
make this constructor explicit.

tjhei commented on include/deal.II/dofs/dof_vector.h:
```diff
@@ -0,0 +1,243 @@
+  /**
+   * Destructor, deleting own data.
+   */
+  ~DoFVector ();
```
Should the destructor be virtual?

tjhei commented:
> Assignment and copy constructor are worth discussing

If you don't want to think about it, you can always make them private for now.

bangerth commented on include/deal.II/dofs/dof_vector.h:
```diff
@@ -0,0 +1,243 @@
+ * Data vectors for finite element discretizations do not make sense
+ * without their DoFHandler. Nevertheless, deal.II has assumed for
+ * decades, that there is a single DoFHandler in a program and that
+ * all vectors refer to it. This class encapsulates both, to begin
```
The assertion isn't quite true -- several tutorial programs, for example, use multiple DoFHandler objects. In any case, I've found that temporal references (like, "the previous behavior of the library was to...") aren't really useful: users only want to know what the behavior is now. If you want to know what changed, then you look into the changelog; how a class or group of classes behaved is of no interest to those who read a class documentation.

bangerth commented on include/deal.II/dofs/dof_vector.h:
```diff
@@ -0,0 +1,243 @@
+ * structures.
+ *
+ * @todo Ensure that both objects are synchronized, that is, the
+ * DoFHandler is not changes without the vector being resized.
```
changes -> changed.

You can do this via a signal.

bangerth commented on include/deal.II/dofs/dof_vector.h:
```diff
@@ -0,0 +1,243 @@
+   * @note This constructor uses the ConstraintMatrix
+   * #no_constraints. The created object thus can output vectors in
+   * distributed format, but writing into vectors will most likely
+   * fail without using a constraint matrix from somewhere else.
```
I don't think I understand what this paragraph means:

* Is `no_constraints` an internal variable or something the user is supposed to understand?

* What does "output vectors in distributed format" mean?

* Why would writing into vectors fail? Do you want to say that doing something like copy-local-to-global will yield the *wrong results* without an appropriate constraints object?

bangerth commented on include/deal.II/dofs/dof_vector.h:
```diff
@@ -0,0 +1,243 @@
+   * order to avoid obscure access errors to the data vector, it
+   * should also be declared `const`.
+   */
+  DoFVector (const DH &dh, const ConstraintMatrix &constraints, const VECTOR &v);
```
The comment says "...and borrow a vector from GrowingVectorMemory". Is this a copy-paste error?

bangerth commented on include/deal.II/dofs/dof_vector.h:
```diff
@@ -0,0 +1,243 @@
+   * distributed format, since the information on constraints is
+   * missing.
+   */
+  DoFVector (const DH &dh, const VECTOR &v);
```
same as above?

bangerth commented on include/deal.II/dofs/dof_vector.h:
```diff
@@ -0,0 +1,243 @@
+   * Modify the vector dimensions to match the current state of the
+   * dof handler.
+   */
+  void sync ();
```
what happens to the content of the vector? Is it just reset to zero? Is it transferred?

bangerth commented on include/deal.II/dofs/dof_vector.h:
```diff
@@ -0,0 +1,243 @@
+  /**
+   * Empty constrainty matrix for constructors without one.
+   */
+  ConstraintMatrix no_constraints;
```
If it remains empty, you could make it `const`. In fact, you could probably make it `static const`.

bangerth commented on include/deal.II/dofs/dof_vector.h:
```diff
@@ -0,0 +1,243 @@
+  /**
+   * The data vector if referencing a vector not owned by this object.
+   */
+  SmartPointer<const VECTOR, DoFVector<DH, VECTOR> > other_data;
```
Am I correct that either `my_data` or `other_data` are always NULL? If so, maybe add a comment.

bangerth commented on include/deal.II/numerics/fe_field_function.h:
```diff
@@ -49,6 +50,11 @@ namespace Functions
+   * @note This function is not intended for assembling a vector or a
+   * matrix on a cell-by-cell basis. Tailored access through FEValues
+   * or similar is recommended then. In such a case, the base class
+   * DoFVector may be helpful.
```
I don't understand this comment. First, it's a class documentation, so what does "This function" refer to. Second, if the base class is useful, then it needs to be visible to the user -- but you use `private` inheritance.

guidokanschat replied:
You should never use `FEFieldFunction` if you have an `FEValues` object, since access through `FEValues::get_function_values()` is much, much faster.

bangerth commented on include/deal.II/dofs/dof_vector.h:
```diff
@@ -0,0 +1,243 @@
+  /**
+   * Access to the DoFHandler
+   */
+  const DH &dof_handler () const;
```
Would you be willing to call it `get_dof_handler()` to be consistent with other places? (And possibly also in the next function: `data()`->`get_data()`?)

bangerth commented on source/dofs/dof_vector.cc:
```diff
@@ -0,0 +1,91 @@
+  void reinit_dof_vector (const DoFHandler<dim, spacedim> &dh,
+                          BlockVector<number> &v)
+  {
+    v.reinit(dh.block_info().global());
```
This assumes a very specific data layout that may or may not be what the user has in mind.

bangerth commented:
I understand the intent to have a way to keep the description of the space (the DoFHandler) and the values of unknowns defined on it (the solution vector) together.

That said, I think we never got there because there are just too many moving parts. People partition their unknowns in ways that are hard to predict and unknowable in a function such as `sync()`. This includes how users want to partition variables into vector blocks, but the same is essentially true for parallel vectors. How do you intend to address this? Right now, you just make a decision (partition so that every FE block corresponds to one vector block) and you don't allow parallel vectors.

The other question I would have is this: With this class, people would have (i) the same member variables as before, i.e., a DoFHandler, a ConstraintMatrix, and possibly (but no longer necessarily) a plain vector; (ii) they would be able to add another member variable, the `DoFVector`, that ties all of these together into one; (iii) but to do anything with it, one currently has to unpack it all again.

Which direction did you intend to evolve this class into? The current extent of it, tying variables together that then immediately need to be accessed individually again, does not seem to provide much in terms of functionality that is any easier than what's currently already there.

guidokanschat commented:
The updated pull request below (conflicts not resolved yet to preserve history of inline comments) addresses all of the comments except for the following: the data layout that users may want to use. While I agree, that very different layouts may be possible, there are some standard layouts, that will cover a lot of applications. Users who have such a layout can use `sync()` for all their vectors and be safe. Others cannot use sync, unless we have a better idea. This does not affect parallel vectors. They are not included, since I do not want to implement anything wrong and my experience is limited. As soon as I understand how they fir into the scheme, they will be added.

Then, the question about the purpose was asked: right now, this class is indeed not useful. Actually, it is not used except as a private base class of `FEFieldFunction`.

The immediately intended application is using them them for `MeshWorker::IntegrationInfo` and similar classes in order to be able to easily use data from previous simulations with different variables.

Later, once the design of this class is stable and agreed on (what about vectors not owned), I suggest to replace the functions in VectorTools in order to make their calls more intuitive.

guidokanschat commented:
Oh well, it outdated the diffs anyway.

bangerth commented on include/deal.II/dofs/dof_vector.h:
```diff
@@ -0,0 +1,272 @@
+ *
+ * This class encapsulates a data vector with the finite element space
+ * information associated with it, such that every function which is
+ * not purely algebraic can assogiate the right finite element
```
assogiate -> associate

bangerth replied:
Did you mean to say "...such that every *vector* which is not purely algebraic..."?

I think I have a hunch what "purely algebraic" is supposed to mean (presumably just an element of R^n, without the connotation that the elements correspond to the nodal vectors of a finite element space) but I think that would be worthwhile making clearer.

bangerth commented on include/deal.II/dofs/dof_vector.h:
```diff
@@ -0,0 +1,272 @@
+ * This class encapsulates a data vector with the finite element space
+ * information associated with it, such that every function which is
+ * not purely algebraic can assogiate the right finite element
+ * function with the data vector. Thus, functions only need a single
```
There are too many uses of the term "function" here. Maybe "Thus, functions in deal.II or user codes"?

bangerth commented on include/deal.II/dofs/dof_vector.h:
```diff
@@ -0,0 +1,272 @@
+
+  /**
+   * The actual data vector, if owned by this object. This pointer
+   * nonzero implies #other_data is zero.
```
grammar. same below.

bangerth commented:
I think I still struggle to see where this class fits in. If I look at where we currently pass around solution vectors, then this is:
* During assembly. Here, we typically really just need the algebraic properties of vectors. In fact, what one assembles is the right hand side vector, which is *not* a nodal vector and so isn't covered by this class. (It would also likely confuse users to use two different classes for solution and right hand side; though it may be a useful tool to educate them about the difference.) On the other hand, one *does* sometimes access the solution vector during assembly, e.g., to preset the values of constrained DoFs.
* During solution. This is purely algebraic. The vector is the appropriate structure, since we make no use of the topology of the finite element space in linear solvers.
* In evaluating the solution. This is where the many functions in VectorTools come in, and this is where I could see a use for this class. On the other hand, many of the functions there take multiple vectors for the same DoFHandler. Using DoFVector would not simplify these calls.

Could you outline concrete cases where you want to go with this class and how this would make things either simpler or easier to understand for users?

guidokanschat commented:
* During assembly. Here we only need the structure of the finite element space, not the linear structure. We also not only need vectors to write in, but also input vectors from previous Newton or time steps, or vectors from an operator splitting approach. Writing a BDF scheme for varying adaptive meshes. Using `DoFVector` in an `AnyData` object, you can automatically compute values in quadrature points of any vector in a consistent way. This is the application I have in mind right now.

* During solution. In a matrix free world, vectors move a step away from R^n and closer to function spaces, since every application of an operator is an integral, which in turn requires to know the function values associated with a vector. For this, we would have to add the interface of the linear structure of `Vector` to `DoFVector`. In return we can test, whether a vector is the right one for a certain operation.

By the way, the use of `DoFVector` obsoletes the difference between compressed and distributed vectors (functions and forms), since it always accesses values on a cell through the `ConstraintMatrix`.

tjhei commented:
I agree with Wolfgang that it might be a good idea to show an example (maybe just stuff it into tests/) how this could be useful. 

Where this could be really neat is when you have more than one DoFHandler and you want to evaluate some coefficient in the assembly loop of a different DoFHandler. I am not quite sure how the interface should look like, but efficient get_function_values() would be awesome (that doesn't use ```point_value```).

tamiko commented:
This class is a special case of a broader idiom we never really followed so far. For example, in my code I usually have an object
```
template <int dim>                                                               
class Discretization : public dealii::Subscriptor                                
{                                                                                
public:                                                                          
  Discretization(const dealii::Triangulation<dim> &triangulation,                
                 const dealii::Mapping<dim> &mapping,                            
                 const dealii::FiniteElement<dim> &finite_element,               
                 const dealii::Quadrature<dim> &quadrature_formula); 
  // ...
};
```
One needs this information always in one place, so a placeholder ```Discretization``` is a very convenient object for passing this information around.

But back to the point: We have an incredibly large number of functions that need a ```Vector``` and its associated ```DoFHandler``` and ```ConstraintMatrix```: Every ```projection``` and ```interpolation``` function does, as well as ```DataOut```, etc. This is definitely a point for defining such an object - at least for bundling ```DoFHandler```, ```ConstraintMatrix``` and ```Vector```.

The only problem I have is that an introduction of ```DoFVector``` creates an unfortunate dichotomy: Which shall be the blessed call interface?
  - provide high level functions acting on the wrapper classes, i.e.,

    ```
    interpolate(const DoFVector<...> &...)
    ```

  - force the break-down to individual components:

    ```
    interpolate(const DoFHandler<...> &...,
                const ConstraintMatrix &...,
                const Vector<...> &)
    ```

Further, the ```DoFVector``` class makes policy decisions with respect to storage that I would like to avoid.

bangerth commented:
One of the fundamental departures in the design of deal.II from many of the previous finite element libraries (since repeated in basically everything that came after deal.II) is that we have a 1:N relationship in essentially all of the semantic dependencies:
* One triangulation can serve many DoFHandlers
* One DoFHandler can serve many solution vectors
* One finite element object can serve many DoFHandlers
* One sparsity pattern can serve many matrices.

This was a conscious design decision: previous libraries typically stored DoF indices directly in the triangulation, and then continued to store solution data (and sometimes local matrices) in the triangulation as well. Such a design only allows for 1:1 relationships, or at the very least makes it more difficult to serve the 1:N pattern. I'm actually quite proud of the fact that we saw this at the time.

The downside is that users not only can, but have to store all of these objects (triangulation, DoFHandler, vectors, sparsity patterns, matrices, ...) separately. It is this downside that @guidokanschat 's patch tries to address, but I will admit that I feel that it takes us away from the flexibility we have. For example, many programs will store multiple solution vectors (think: time dependent or nonlinear problems), and in that case one would either have to
* store multiple `DoFVector` objects all of which simply refer to the common DoFHandler and constraint object
* store multiple vectors as one would currently, but then when calling a function such as `interpolate` create a `DoFVector` object on the fly:
```
  interpolate (DoFVector(dof_handler, constraints, solution));
```

The first of these alternatives provides no savings in the number of member variables, just a bit of confusion over who actually owns what. The second doesn't make writing code simpler. So what I miss is an example of where the new class provides a significant advantage when consider *large* and *complex* programs (for 300-line tutorial programs, complexity is not an issue).

bangerth commented:
There's been no progress in 4.5 months, so I'll close this issue for now. We can always re-open it if there is activity.
